0002-snowflake-ids.md

```diff
@@ -0,0 +1,71 @@
+# Snowflake IDs
+
+<!-- Fill these out once the PR has been created. Remember to remove comments once they're no longer applicable. -->
+
+- Status: Pending
+- Start Date: 2023-04-13
+- RFC PR: [#2](https://github.com/loki-chat/rfcs/pull/2)
+- Tracking issue: N/A <!-- When this RFC is made active, add link in the form of [<repo name>#0](https://github.com/loki-chat/<repo name>/issues/0) if applicable. -->
+- Supercedes: N/A
+
+## Summary
+
+This RFC specifies a singular ID format to be used across Loki, for things like messages and users, called snowflakes. These are proven to work well for avoiding ID collisions, even across distributed systems, by their usage at companies like Twitter and Discord.
+
+## Basic example
+
+The ID 36889232862682412 may be deconstructed as follows:
+
+| Timestamp in milliseconds, UTC (42 bits)   | Worker ID (10 bits)  | Counter (12 bits) |
+| ------------------------------------------ | -------------------- | ----------------- |
+| 000000001000001100011001111111100100010100 | 1110010110           | 111100010110      |
+| 8798075156                                 | 918                  | 3862              |
+
+We can get the Unix time of the ID by adding the Unix time of our epoch, which is midnight, January 1st, 2023. 8798075156 + 1672531200000 gives us 1681329275156. This gives us a correct time of April 12th, 2023, at 19:54:35.156 (7:54 PM), in UTC.
+
+## Motivation
+
+Standardizing on a single ID format ensures consistency, it means that existing ID-related code is more easily reusable, and it prevents us from reinventing the wheel in separate parts of the codebase.
+
+An ID's job being to identify a certain thing, and only that thing, it should be unique. This should ideally also be scalable, such that two workers have zero chance of assigning two separate things the same ID and causing problems.
+
+Finally, an ID beginning with the timestamp can easily be sorted by, and provide a reference point to get messages before and after it.
+
+## Detailed design
+
+Snowflakes are stored in 64-bit unsigned integers. The client may only depend on the timestamp, as the other parts may be changed without warning.
+
+### Timestamp
+
+Bits 63 through 22 are used for the timestamp. The epoch for this timestamp is midnight, January 1st, 2023. The Unix timestamp for this epoch is 1672531200000. This means that to convert to Unix time, the timestamp of the epoch should be added, and vice versa. This has space for times up to May 15th, 2162.
+
+The timestamp of an ID may never be less than that of an ID generated earlier with the same worker ID. IDs across separate workers should be sufficiently close for the purpose of sorting, ideally less than 1 second.
+
+In cases where only signed integers are available, the first bit of the timestamp may be ignored by the client, to account for the sign bit.
```

Comment:
I think this should be removed since this effectively means there might be id collision for when the last bit is needed but ignored, the only time signed integers would be required is for postgres which the backend will deal without anything else needing to think about it.

Reply:
The thing is that if a client doesn't support signed integers, it will likely just run with unsigned ones either way. 41 bits will still work until September 2092, giving us plenty of time to require using unsigned integers and preserving the bit if necessary.

Reply:
In what cases would the client ever not support both signed and unsigned integers? There is just no point in thinking about such a possibility

Reply:
It could be the case that the language used simply only has signed integers. I do not personally use any such language where this particularly is a problem. However, I do not think that's an excuse to not account for people who do. Besides, it's not a problem for anyone else.

I looked it up, and Haskell does not make the distinction. With GHC on x86-64, [its `Int` type is 64-bit signed](https://wiki.haskell.org/Converting_numbers#Converting_from_and_between_integral_types_.28integer-like_types.29).

Reply:
I do think it's a perfectly fine excuse when we will be building our own client which does have them, languages which decide to not have unsigned integers will simply just not correctly print out the ID anymore, either way this issues will only occur after September 2092 and so making them unsigned now or then is irrelevant

Reply:
Also
https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html
"A [Word](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html#t:Word) is an unsigned integral type, with the same size as [Int](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Int.html#t:Int)."

Reply:
> I do think it's a perfectly fine excuse when we will be building our own client which does have them, languages which decide to not have unsigned integers will simply just not correctly print out the ID anymore, either way this issues will only occur after September 2092 and so making them unsigned now or then is irrelevant

@Coca162 I'm also thinking about bots, which might be implemented in anything. The point is just to say "Hey, if you must use signed integers, that's fine, don't sweat it". It makes no meaningful difference in behavior, it just means that affected developers don't have to worry until way off in the future, assuming we survive until then.

> Also https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html "A [Word](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html#t:Word) is an unsigned integral type, with the same size as [Int](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Int.html#t:Int)."

Ah, I see. Point about other languages remains though.

Reply:
> I'm also thinking about bots, which might be implemented in anything.

A large majority of languages people want to write in also have unsigned integers so the point still fails to stand. But if I were to entertain the argument then the timestamp should just simply be 41 bits long with in the future an extension to i128. Having this condition is just bound to cause trouble.

Reply:
> A large majority of languages people want to write in also have unsigned integers so the point still fails to stand. But if I were to entertain the argument then the timestamp should just simply be 41 bits long with in the future an extension to i128. Having this condition is just bound to cause trouble.

I am not rejecting a later migration if this might ever cause an issue. If the issue I'm suggesting ends up being relevant, we might migrate. If not, then we avoid migrating for a lot longer. I consider that a win-win.

Given that, it doesn't cause problems for the "large majority of languages", ie. it's not a big deal. It does however accommodate that minority that still does exist and is not necessarily unused.

Reply:
It's not a question of if this causes it a issue, it WILL cause a issue. You cannot simply "ignore" (or whatever that means) part of ID and then assume it's unique. Either accept the loss that your ID is going to display weirdly or turn the ID into a i64 and have 41 bits for the timestamp.


> It does however accommodate that minority that still does exist and is not necessarily unused.

Look I said that because I was curious to see any languages which people might use don't have them, because I can't think of any and neither can you it seems.

Reply:
> It could be the case that the language used simply only has signed integers. I do not personally use any such language where this particularly is a problem. However, I do not think that's an excuse to not account for people who do. Besides, it's not a problem for anyone else.
> 
> I looked it up, and Haskell does not make the distinction. With GHC on x86-64, [its `Int` type is 64-bit signed](https://wiki.haskell.org/Converting_numbers#Converting_from_and_between_integral_types_.28integer-like_types.29).

Java is a trivial example of a language that doesn't support unsigned integers types. However, even then that is not a problem, because:
- the bits stay the same, therefore if you do bitshift calculations to retrieve the individual parts of the timestamp, they will stay intact no matter the signedness of the type used. This signedness is only relevant for doing operations on it as a number, AFAIK.
- like in any other language, if the sign is a problem, there's a solution, however ugly it may be: in Java, there is the `BigDecimal` type in the standard library, and in Haskell there is the `Word` type as suggested. But that's pretty much irrelevant when considering my first point.

Reply:
> It's not a question of if this causes it a issue, it WILL cause a issue. You cannot simply "ignore" (whatever that means) part of ID and then assume it's unique. Either accept the loss that your ID is going to display weirdly or turn the ID into a i64 and have 41 bits for the timestamp.
> 
> > It does however accommodate that minority that still does exist and is not necessarily unused.
> 
> Look I said that because I was curious to see any languages which people might use don't have them, because I can't think of any and neither can you it seems.

@Coca162 I don't know any people from Uganda but that doesn't mean they don't exist.

Here's the deal though. For two identical timestamps to be generated using this method, it will take almost years. Even if you stretch it, modern computing hasn't even existed for that long. For this to be a problem, we will need to reach almost to 2093 without a single change to the ID format being made whatsoever. That is an unrealistic assumption to say the least.

> the bits stay the same, therefore if you do bitshift calculations to retrieve the individual parts of the timestamp, they will stay intact no matter the signedness of the type used. This signedness is only relevant for doing operations on it as a number, AFAIK.

@Speykious It is relevant to being able to sort by ID directly. Like I said above though, it's unlikely to become an issue.

Reply:
> without a single change to the ID format being made whatsoever

This is very likely possible if the chat app continues running, migration is a pain in the ass.

> That is an unrealistic assumption to say the least.

You have to keep in mind that the RFC specifically says you can do this, there is no reason to state this if it is not thinking about the long term. If we don't care about 2093 then the line should simply just be removed.

Reply:
That's no problem for the server, since that's written in Rust. For an alternative client, I could see the argument, but at the same time, a client won't have to deal with as many IDs as the server, so using a custom sorting algorithm (by comparing the timestamp part and then the other parts) is no big deal there.

Reply:
Also, I do not understand the practical applications of sorting due to the worker IDs, are there any use cases?

Reply:
> This is very likely possible if the chat app continues running, migration is a pain in the ass.

@Coca162 So then we ought to throw out this format and start work on one that can last a lot longer? That seems like an issue no matter what we do.

> You have to keep in mind that the RFC specifically says you can do this, there is no reason to state this if it is not thinking about the long term. If we don't care about 2093 then the line should simply just be removed.

@Coca162 It is not thinking about the long term. It is about saying right now that doing that is okay.

> That's no problem for the server, since that's written in Rust. For an alternative client, I could see the argument, but at the same time, a client won't have to deal with as many IDs as the server, so using a custom sorting algorithm (by comparing the timestamp part and then the other parts) is no big deal there.

@Speykious That is true.

> Also, I do not understand the practical applications of sorting due to the worker IDs, are there any use cases?

@Coca162 I'm not sure what you mean by this.

> I can imagine some kind of sorting being needed client-side for when fetching a bulk of old messages, but maybe it's not needed.

@Speykious The client might reasonably assume that messages are returned in the correct order. Or it might not. That depends on how our API is defined.

Reply:
Because JS does not have integers at all, we could go the route of discord and have it be passed around as a string in some languages.